From 156b521d6181d5447895cf17b0436ff83b053629 Mon Sep 17 00:00:00 2001
From: wh <wanghang@rock-chips.com>
Date: Wed, 9 Dec 2020 11:45:13 +0800
Subject: [PATCH 3/5] Gallery2: fix can not save pic after edit to extend
 storage

1. Based on the current test results, it seems the extend storage write state:
   udisk ntfs: /mnt/media_rw/; udisk fat32: /storage/ ; sdcard(maybe fat32): /storage/;
2. this commit also need combine with another commit in packages/providers/MediaProvider

Signed-off-by: wh <wanghang@rock-chips.com>
Change-Id: I28dbbe6b4e822c38efa09dccddfbd011592cf4b7
---
 AndroidManifest.xml                                  |  1 +
 src/com/android/gallery3d/data/DataManager.java      | 12 ++++++++++++
 .../gallery3d/filtershow/tools/SaveImage.java        |  4 +++-
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index a1298ab48..4fd1fc7dc 100644
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -32,6 +32,7 @@
     <uses-permission android:name="android.permission.WRITE_MEDIA_STORAGE" />
     <uses-permission android:name="android.permission.WRITE_SETTINGS" />
     <uses-permission android:name="android.permission.WRITE_SYNC_SETTINGS" />
+    <uses-permission android:name="android.permission.MANAGE_EXTERNAL_STORAGE"/>
     <uses-permission android:name="com.android.gallery3d.permission.GALLERY_PROVIDER" />
 
     <supports-screens android:smallScreens="false"
diff --git a/src/com/android/gallery3d/data/DataManager.java b/src/com/android/gallery3d/data/DataManager.java
index 3cf278712..40ea0e8c4 100644
--- a/src/com/android/gallery3d/data/DataManager.java
+++ b/src/com/android/gallery3d/data/DataManager.java
@@ -395,6 +395,18 @@ public class DataManager implements StitchingChangeListener {
         return path;
     }
 
+    public static String convertStorageToMntCanWrite(String path) {
+        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.O_MR1
+                && isExtendStorage(path)) {
+            String[] pathSplit = path.split("/");
+            if (null != pathSplit && pathSplit.length > 3
+                && new File(MNT_PATH + "/" + pathSplit[2]).canWrite()) {
+                return path.replaceFirst(STORAGE_REGEX, MNT_PATH);
+            }
+        }
+        return path;
+    }
+
     private static boolean isExtendStorage(String path) {
         return null != path && path.startsWith("/storage/")
                 && !path.startsWith("/storage/emulated/");
diff --git a/src/com/android/gallery3d/filtershow/tools/SaveImage.java b/src/com/android/gallery3d/filtershow/tools/SaveImage.java
index 8c2b49a86..e48f9f3a8 100644
--- a/src/com/android/gallery3d/filtershow/tools/SaveImage.java
+++ b/src/com/android/gallery3d/filtershow/tools/SaveImage.java
@@ -32,6 +32,7 @@ import android.widget.Toast;
 
 import com.android.gallery3d.R;
 import com.android.gallery3d.common.Utils;
+import com.android.gallery3d.data.DataManager;
 import com.android.gallery3d.exif.ExifInterface;
 import com.android.gallery3d.filtershow.FilterShowActivity;
 import com.android.gallery3d.filtershow.cache.ImageLoader;
@@ -268,7 +269,8 @@ public class SaveImage {
         boolean ret = false;
         OutputStream s = null;
         try {
-            s = exif.getExifWriterStream(file.getAbsolutePath());
+            s = exif.getExifWriterStream(
+                    DataManager.convertStorageToMntCanWrite(file.getAbsolutePath()));
             image.compress(Bitmap.CompressFormat.JPEG,
                     (jpegCompressQuality > 0) ? jpegCompressQuality : 1, s);
             s.flush();
-- 
2.35.1

